rust/feature-flags/src/flag_matching.rs

```diff
@@ -32,73 +166,288 @@ pub struct Person {
+                    .iter()
+                    .filter_map(|flag| flag.get_group_type_index())
+                    .collect(),
+            )
+            .await
+        {
+            error_while_computing_flags = true;
+            error!("Error fetching properties: {:?}", e);
+            return FlagsResponse {
+                error_while_computing_flags,
+                feature_flags: result,
+            };
+        }
+
+        // Step 3: Evaluate each flag
+        for flag in feature_flags.flags {
+            if !flag.active || flag.deleted {
+                continue;
+            }
+
+            match self.get_match(&flag).await {
+                Ok(flag_match) => {
+                    let flag_value = if flag_match.matches {
+                        match flag_match.variant {
+                            Some(variant) => FlagValue::String(variant),
+                            None => FlagValue::Boolean(true),
+                        }
+                    } else {
+                        FlagValue::Boolean(false)
+                    };
+                    result.insert(flag.key.clone(), flag_value);
+                }
+                Err(e) => {
+                    error_while_computing_flags = true;
+                    error!(
+                        "Error evaluating feature flag '{}' for distinct_id '{}': {:?}",
+                        flag.key, self.distinct_id, e
+                    );
+                }
+            }
+        }
+
+        FlagsResponse {
+            error_while_computing_flags,
+            feature_flags: result,
+        }
+    }
+
+    async fn fetch_and_cache_properties(
+        &self,
+        group_type_indexes: &HashSet<GroupTypeIndex>,
+    ) -> Result<(), FlagError> {
+        if self.database_client.is_none() {
+            error!("Database client is None");
+            return Err(FlagError::DatabaseUnavailable);
+        }
+
+        let mut conn = self
+            .database_client
+            .as_ref()
+            .expect("client should exist here")
+            .get_connection()
+            .await?;
+
+        let query = r#"
+            SELECT 
+                (SELECT "posthog_person"."properties"
+                 FROM "posthog_person"
+                 INNER JOIN "posthog_persondistinctid" 
+                 ON ("posthog_person"."id" = "posthog_persondistinctid"."person_id")
+                 WHERE ("posthog_persondistinctid"."distinct_id" = $1
+                        AND "posthog_persondistinctid"."team_id" = $2
+                        AND "posthog_person"."team_id" = $2)
+                 LIMIT 1) as person_properties,
+                
+                (SELECT json_object_agg("posthog_group"."group_type_index", "posthog_group"."group_properties")
+                 FROM "posthog_group"
+                 WHERE ("posthog_group"."team_id" = $2
+                        AND "posthog_group"."group_type_index" = ANY($3))) as group_properties
+        "#;
```

Comment:
I tried my best to optimize this query by inlining as much as I could (instead of creating CTEs), but the rest of the optimization would rely on the tables being indexed.  This feels good enough for now; the new calling patterns should speed up this service.